app/src/main/java/org/ea/sqrl/utils/PasswordStrengthMeter.java

```diff
@@ -0,0 +1,230 @@
+
+/**
+ * Calculates and displays a password strength rating.
+ * Expects the password_strength_meter.xml to be present in the calling layout
+ *
+ * @author Alexander Hauser (alexhauser)
+ */
+public class PasswordStrengthMeter {
+
+    private final int PW_MIN_LENGTH = 8;
+    private final int STRENGTH_POINTS_MIN_MEDIUM = 5;
+    private final int STRENGTH_POINTS_MIN_GOOD = 9;
+    private final Context mContext;
+    private ViewGroup mPassStrengthLayout;
+    private CalcPasswordStrengthAsyncTask mLastCalcTask;
+
+
+    /**
+     * Creates a PasswordStrengthMeter object.
+     *
+     * @param context The context of the caller.
+     */
+    public PasswordStrengthMeter(Context context) {
+
+        mContext = context;
+    }
+
+    /**
+     * Registers a TextChangedListener for the given password field and calculates
+     * and displays the password strength each time the listener fires.
+     *
+     * @param passwordField The password field to be observed.
+     * @param passStrengthLayout A ViewGroup containing the password_strength_meter.xml layout.
+     */
+    public void register(EditText passwordField, ViewGroup passStrengthLayout) {
+
+        mPassStrengthLayout = passStrengthLayout;
+        passwordField.addTextChangedListener(new TextWatcher() {
+
+            public void afterTextChanged(Editable s) {
+                String password = s.toString();
+
+                if (mLastCalcTask != null) mLastCalcTask.cancel(true);
+                mLastCalcTask = new CalcPasswordStrengthAsyncTask();
+                mLastCalcTask.execute(password);
+            }
+
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
+            public void onTextChanged(CharSequence s, int start, int before, int count) {}
+        });
+    }
+
+
+    class CalcPasswordStrengthAsyncTask extends AsyncTask<String, Void, PasswordStrengthResult> {
+
+        @Override
+        protected PasswordStrengthResult doInBackground(String... args) {
+
+            // This will avoid updating the UI too often on quick password input
+            for (int i=0; i<10; i++) {
+                try { Thread.sleep(5); } catch(Exception e) {}
+                if (isCancelled()) return null;
+            }
+
+            PasswordStrengthResult result = new PasswordStrengthResult();
+            CharacterClass characterClass;
+            String password = args[0];
+
+            result.PasswordLength = password.length();
+            if (result.PasswordLength <= 5) result.StrengthPoints = 0;
+            else if (result.PasswordLength <= 8) result.StrengthPoints = 1;
+            else if (result.PasswordLength <= 10) result.StrengthPoints = 2;
+            else if (result.PasswordLength <= 13) result.StrengthPoints = 5;
+            else if (result.PasswordLength <= 16) result.StrengthPoints = 10;
+            else result.StrengthPoints = 30;
+
+            for (int i=0; i < result.PasswordLength; i++) {
```

Comment:
A micro-optimization would be to set booleans for the different classes and only check for the ones that you are missing and if all are set you don't loop through the rest.

Reply:
I wanted to be able to support displaying class counts, just as Steve does on his password haystack page, that's why I used ints. But I guess for our purposes, it wouldn't even be useful. I'll refactor the code to use booleans.